### PR TITLE
feat: mode for plugin

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

The idea is from @Atinux , use `.server.js` or `.client.js` to automatically detect server/client side plugin.

1. introduce `mode: 'client|server'` in plugin options
2. deprecate `ssr` in plugin options and use `mode: 'client|server'` instead, for now auto adapt `ssr: false` to `mode: 'client'` and show deprecating log.
3. apply `server/client` plugin on file ext like: `server.js` or `client.js` (**specified options has higher priority**)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/builder/src/builder.js [hunk 1]
```diff
@@ -110,9 +110,18 @@ export default class Builder {
           /[^a-zA-Z?\d\s:]/g,
           ''
         )
+
+        // TODO: remove deprecated check in Nuxt 3
+        if (p.ssr === false) {
+          p.mode = 'client'
+          consola.warn(`ssr in plugin options has been deprecated, please use mode instead`)
```
Comment (pi0) [inline packages/builder/src/builder.js]: This warning could be really annoying as modules should still prefer using it for 2.x compat. I suggest doing warn on N3 upwards.

File: packages/builder/src/builder.js [hunk 2]
```diff
@@ -224,6 +240,8 @@ export default class Builder {
     // Plugins
     this.plugins = Array.from(this.normalizePlugins())
 
+    await this.resolvePlugins()
```
Comment (pi0) [inline packages/builder/src/builder.js]: Why we should resolve plugins after build?

File: packages/builder/src/builder.js [hunk 3]
```diff
@@ -110,9 +110,18 @@ export default class Builder {
           /[^a-zA-Z?\d\s:]/g,
           ''
         )
+
+        // TODO: remove deprecated check in Nuxt 3
+        if (p.ssr === false) {
+          p.mode = 'client'
+          consola.warn(`ssr in plugin options has been deprecated, please use mode instead`)
+        } else if (!['client', 'server'].includes(p.mode)) {
+          p.mode = 'all'
```
Comment (pi0) [inline packages/builder/src/builder.js]: A warning about invalid mode can be useful here. Users may accidentally set things like `clien`, ... that we silently translate it to `all`.

Comment (clarkdo) [reply packages/builder/src/builder.js]: `all` is not a options that user need to use, it's just a default condition for both client and server.

`client/server` are the only valid value that should be configured in `mode`, if we want to add warning here, we need to another `else if` for `p.mode === undefined`, if it's ok for you, I can add.

Comment (pi0) [reply packages/builder/src/builder.js]: True, `all` is not what people should use by default. But invalid values (typos) are silently ignored.

Comment (clarkdo) [reply packages/builder/src/builder.js]: OK, I'll add the message.

BTW, this can be also added in eslint-plguin

Comment (manniL) [reply packages/builder/src/builder.js]: :+1: for that idea

Comment (pi0) [reply packages/builder/src/builder.js]: Yes. Supporting in ESLint plugin would be awesome.) But we also need to have proper CLI warns too.

Comment (clarkdo) [reply packages/builder/src/builder.js]: Warn msg added

**Review by pi0 — APPROVED** [review]

**Comment by clarkdo** [comment]
Will add a pr for doc

**Comment by Atinux** [comment]
This is awesome @clarkdo

**Comment by CurtisBelt** [comment]
@clarkdo Hello, I am using nuxt-edge while project is in development, and after updating to a more recent commit, none of my plugins load when using `nuxt start`, but they work fine when using `nuxt dev`.

I did some trial and error, moving backwards through available `nuxt-edge` versions, and determined that the problem starts after this feature was merged. The issue still exists as of `2.4.0-25768669.8786ff73` (8786ff731767425bb0c1c72af6d8a8895155acf5).

I did not create a reduced test since I am curious if you will simply reproduce bug right away when using `nuxt start`. If you are unable to reproduce, I am happy to do more research. Thank you!

**Comment by clarkdo** [comment]
@CurtisBelt I’ll look into it, thank you for the reporting 🙏

**Comment by clarkdo** [comment]
Hi @CurtisBelt Is your plugin specified ssr:false? If yes, it won’t be loaded in server side

**Comment by manniL** [comment]
@clarkdo I experience the same on latest nuxt edge for https://github.com/Developmint/developmint.de

**Comment by manniL** [comment]
@clarkdo You need `modern: 'server'` set to trigger the error.

nuxt.config.js
```js
  GNU nano 2.9.3                    nuxt.config.js                              

module.exports = {
  modern: 'server',
  plugins: [
{ src:    '~/plugins/test', ssr: false}
  ]
}
```

plugins/test.js
```js
window.test = true
```

**Comment by manniL** [comment]
Update: fix @ #4659

**Pull request merged**